HIP/hip-0000-introduction-of-blocks.md

```diff
@@ -0,0 +1,199 @@
+
+## Rationale
+
+Hedera must have a single consistent answer to what a block number is, the hash and which transactions are included in that block number. This is required for two reasons:
+
+1. The Smart Contracts Service to have the context of a `number` and the `hash` of the current block while running the EVM bytecode.
+2. Mirror Nodes to have the information in order to implement standard JSON RPC endpoints and be able to answer queries such as:
+    1. Getting current block number
+    2. Getting block by hash/number
+    3. Getting a list of transactions for a block by number/hash
+    4. Getting logs based on block number filtering
+
+**Design Goal #1 Minimize Changes**
+
+The Block concept should fit naturally into the existing processes, mechanisms and state updates. It must keep the same responsibilities between consensus, services and mirror nodes.
+
+**Design Goal #2 Lightweight**
+
+The Block concept must not add a lot of complexity and performance overhead to the processing of transactions. It must have minimal impact on the TPS of the network.
+
+Based on the described design goals above, the outlined specification defines that block properties are to be computed and populated at different points in time and by different components of the network, based on their responsibility.
+
+## Specification
+
+**Definitions**
+
+- `block` → `Record file` containing all `Record Stream Objects` grouped in a 2-second time frame.
+- `block number` → consecutive number of the `Record file` that is being incremented by `1` for every new `Record file`. This value can be initially bootstrapped through `Mirror Nodes` and after that maintained by services nodes.
+- `block hash` → The `32 byte keccak256 running hash` of the `Record file`. That is the running hash of the last `Record Stream Object` from the previous `Record file`.
+- `block timestamp` → Instant of consensus timestamp of the first `transaction`/`Record Stream Object` in the `Record file`.
+
+### Platform
+
+Adapt `TimestampStreamFileWriter` to include `blockNumber` in the `Record File`. Introduce a new field `firstConsensusTimeInCurrentFile` to be used as a marker when to start a new Record file. Use the field `lastConsensusTimestamp` to keep track of the last-seen consensus timestamp that was processed. In this way, we can ensure that we have at least `1000ns`  difference between the last processed transaction before a new file is created. In this way if we have a `parent transaction` with at least one `child precompile transaction` they all will be included in the same `block`/`Record file`. Otherwise, we might have a corner case where a `parent transaction` is included in one block, and its `child precompile transaction` falls into the next block since it will increase the `consensus timestamp` with `1ns`. Therefore, the algorithm for checking whether to start a new file would be the following:
```

Comment:
What is the unit of `firstConsensusTime` and `lastConsensusTimestamp`?  seconds, milli?  micro?  nano? The math implies nano but we should be explicit about it.